util/topgen.py

```diff
@@ -1164,40 +1164,48 @@ def render_template(out_name_tpl, out_dir, **other_info):
 
         # C Header + C File + Clang-format file
 
-        # 'clang-format' -> 'sw/autogen/.clang-format'
-        cformat_tplpath = tpl_path / 'clang-format'
-        cformat_dir = out_path / 'sw/autogen'
-        cformat_dir.mkdir(parents=True, exist_ok=True)
-        cformat_path = cformat_dir / '.clang-format'
-        cformat_path.write_text(cformat_tplpath.read_text())
-
-        # 'top_{topname}.h.tpl' -> 'sw/autogen/top_{topname}.h'
-        cheader_path = render_template('top_%s.h',
-                                       'sw/autogen',
-                                       helper=c_helper)
-
-        # Save the relative header path into `c_gen_info`
-        rel_header_path = cheader_path.relative_to(SRCTREE_TOP)
-        c_helper.header_path = str(rel_header_path)
-
-        # 'top_{topname}.c.tpl' -> 'sw/autogen/top_{topname}.c'
-        render_template('top_%s.c', 'sw/autogen', helper=c_helper)
-
-        # 'top_{topname}_memory.ld.tpl' -> 'sw/autogen/top_{topname}_memory.ld'
-        render_template('top_%s_memory.ld', 'sw/autogen')
-
-        # 'top_{topname}_memory.h.tpl' -> 'sw/autogen/top_{topname}_memory.h'
-        memory_cheader_path = render_template('top_%s_memory.h', 'sw/autogen')
-
-        # Fix the C header guards, which will have the wrong name
-        subprocess.run(["util/fix_include_guard.py",
-                        str(cheader_path),
-                        str(memory_cheader_path)],
-                       universal_newlines=True,
-                       stdout=subprocess.DEVNULL,
-                       stderr=subprocess.DEVNULL,
-                       check=True,
-                       cwd=str(SRCTREE_TOP))  # yapf: disable
+        # Since SW does not use FuseSoC and instead expects those files always
+        # to be in hw/top_{topname}/sw/autogen, we currently create these files
+        # twice:
+        # - Once under out_path/sw/autogen
+        # - Once under hw/top_{topname}/sw/autogen
```

Comment:
Could you explain this a little bit more? Normally isn't `out_path` same to `hw/top_{topname}`?

Reply:
Normally (for top_earlgrey) this is the case. However, if we run topgen at build time through FuseSoC this is not true: `out_path` is then somewhere in `build/..` but software still expects the files in to be in `hw/top_{topname}`. I consider this as a temporary solution.